linker/shared/src/main/scala/org/scalajs/linker/backend/webassembly/FunctionBuilder.scala

```diff
@@ -0,0 +1,406 @@
+/*
+ * Scala.js (https://www.scala-js.org/)
+ *
+ * Copyright EPFL.
+ *
+ * Licensed under Apache License 2.0
+ * (https://www.apache.org/licenses/LICENSE-2.0).
+ *
+ * See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.
+ */
+
+package org.scalajs.linker.backend.webassembly
+
+import scala.collection.mutable
+
+import org.scalajs.ir.{OriginalName, Position}
+
+import Instructions._
+import Identitities._
+import Modules._
+import Types._
+
+final class FunctionBuilder(
+    moduleBuilder: ModuleBuilder,
+    val functionID: FunctionID,
+    val functionOriginalName: OriginalName,
+    functionPos: Position
+) {
+  import FunctionBuilder._
+
+  private var labelIdx = 0
+
+  private val params = mutable.ListBuffer.empty[Local]
+  private val locals = mutable.ListBuffer.empty[Local]
+  private var resultTypes: List[Type] = Nil
+
+  private var specialFunctionType: Option[TypeID] = None
+
+  /** The instructions buffer. */
+  private val instrs: mutable.ListBuffer[Instr] = mutable.ListBuffer.empty
+
+  def setFunctionType(typeID: TypeID): Unit =
+    specialFunctionType = Some(typeID)
+
+  def setResultTypes(tpes: List[Type]): Unit =
+    resultTypes = tpes
+
+  def setResultType(tpe: Type): Unit =
+    setResultTypes(tpe :: Nil)
+
+  def addParam(originalName: OriginalName, tpe: Type): LocalID = {
+    val id = new ParamIDImpl(params.size, originalName)
+    params += Local(id, originalName, tpe)
+    id
+  }
+
+  def addParam(name: String, tpe: Type): LocalID =
+    addParam(OriginalName(name), tpe)
+
+  def genLabel(): LabelID = {
+    val label = new LabelIDImpl(labelIdx)
+    labelIdx += 1
+    label
+  }
+
+  def addLocal(originalName: OriginalName, tpe: Type): LocalID = {
+    val id = new LocalIDImpl(locals.size, originalName)
+    locals += Local(id, originalName, tpe)
+    id
+  }
+
+  def addLocal(name: String, tpe: Type): LocalID =
+    addLocal(OriginalName(name), tpe)
+
+  // Instructions
+
+  def +=(instr: Instr): Unit =
+    instrs += instr
+
+  def ++=(instrs: Iterable[Instr]): Unit =
+    this.instrs ++= instrs
+
+  def markCurrentInstructionIndex(): InstructionIndex =
+    new InstructionIndex(instrs.size)
+
+  def insert(index: InstructionIndex, instr: Instr): Unit =
+    instrs.insert(index.value, instr)
+
+  // Helpers to build structured control flow
+
+  def sigToBlockType(sig: FunctionType): BlockType = sig match {
+    case FunctionType(Nil, Nil) =>
+      BlockType.ValueType()
+    case FunctionType(Nil, resultType :: Nil) =>
+      BlockType.ValueType(resultType)
+    case _ =>
+      BlockType.FunctionType(moduleBuilder.functionTypeToTypeID(sig))
+  }
+
+  def ifThenElse(blockType: BlockType)(thenp: => Unit)(elsep: => Unit): Unit = {
```

Comment:
The latest commit uses a typeclass instead. There's still an overload for the empty block type, and `switch` still needs its overloads because it wants `FunctionType`s. Is it better?